File: BackEnd/PHP/laravelshop/app/Http/Controllers/Api/V1/Friend/FriendController.php
<?php

namespace App\Http\Controllers\Api\V1\Friend;

use App\Http\Controllers\Api\V1\ApiController;
use App\Modules\Friend\Services\FriendService;
use Illuminate\Http\Request;

class FriendController extends ApiController
{
    public function __construct(private readonly FriendService $service)
    {
    }

    public function request(Request $request, int $userId)
    {
        return $this->success($this->service->request($request->user(), $userId), [], 201);
    }

    public function accept(Request $request, int $friendshipId)
    {
        return $this->success($this->service->accept($request->user(), $friendshipId));
    }

    public function reject(Request $request, int $friendshipId)
    {
        return $this->success($this->service->reject($request->user(), $friendshipId));
    }

    public function index(Request $request)
    {
        return $this->success($this->service->list($request->user(), (int) $request->query('page', 1), (int) $request->query('limit', 20)));
    }

    public function received(Request $request)
    {
        return $this->success($this->service->received($request->user(), (int) $request->query('page', 1), (int) $request->query('limit', 20)));
    }

    public function sent(Request $request)
    {
        return $this->success($this->service->sent($request->user(), (int) $request->query('page', 1), (int) $request->query('limit', 20)));
    }

    public function feed(Request $request)
    {
        return $this->success($this->service->feed($request->user(), (int) $request->query('page', 1), (int) $request->query('limit', 20)));
    }

    public function block(Request $request, int $userId)
    {
        return $this->success($this->service->block($request->user(), $userId), [], 201);
    }

    public function unblock(Request $request, int $userId)
    {
        return $this->success($this->service->unblock($request->user(), $userId));
    }

    public function destroy(Request $request, int $userId)
    {
        return $this->success($this->service->remove($request->user(), $userId));
    }
}

File: BackEnd/Java/java-spring-maven-jpa-postgresql/src/main/java/com/pbshop/java/spring/maven/jpa/postgresql/friend/FriendController.java
package com.pbshop.java.spring.maven.jpa.postgresql.friend;

import java.util.Map;

import jakarta.validation.Valid;

import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.pbshop.java.spring.maven.jpa.postgresql.auth.security.AuthenticatedUserPrincipal;
import com.pbshop.java.spring.maven.jpa.postgresql.common.api.ApiResponse;
import com.pbshop.java.spring.maven.jpa.postgresql.friend.dto.FriendDtos.BlockUserRequest;
import com.pbshop.java.spring.maven.jpa.postgresql.friend.dto.FriendDtos.RequestFriendRequest;

@RestController
@RequestMapping("${pbshop.api.base-path:/api/v1}/friends")
public class FriendController {
    private final FriendService friendService;
    public FriendController(FriendService friendService) { this.friendService = friendService; }

    @PostMapping("/requests")
    public ApiResponse<Map<String, Object>> request(@AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @Valid @RequestBody RequestFriendRequest request) {
        return ApiResponse.success(friendService.requestFriend(principal, request.userId()));
    }
    @PostMapping("/requests/{id}/accept")
    public ApiResponse<Map<String, Object>> accept(@AuthenticationPrincipal AuthenticatedUserPrincipal principal, @PathVariable Long id) {
        return ApiResponse.success(friendService.accept(principal, id));
    }
    @PostMapping("/requests/{id}/reject")
    public ApiResponse<Map<String, Object>> reject(@AuthenticationPrincipal AuthenticatedUserPrincipal principal, @PathVariable Long id) {
        return ApiResponse.success(friendService.reject(principal, id));
    }
    @GetMapping
    public ApiResponse<Map<String, Object>> list(@AuthenticationPrincipal AuthenticatedUserPrincipal principal) {
        return ApiResponse.success(friendService.getFriends(principal));
    }
    @GetMapping("/requests/received")
    public ApiResponse<Map<String, Object>> received(@AuthenticationPrincipal AuthenticatedUserPrincipal principal) {
        return ApiResponse.success(friendService.getReceived(principal));
    }
    @GetMapping("/requests/sent")
    public ApiResponse<Map<String, Object>> sent(@AuthenticationPrincipal AuthenticatedUserPrincipal principal) {
        return ApiResponse.success(friendService.getSent(principal));
    }
    @GetMapping("/feed")
    public ApiResponse<Map<String, Object>> feed(@AuthenticationPrincipal AuthenticatedUserPrincipal principal) {
        return ApiResponse.success(friendService.getFeed(principal));
    }
    @PostMapping("/blocks")
    public ApiResponse<Map<String, Object>> block(@AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @Valid @RequestBody BlockUserRequest request) {
        return ApiResponse.success(friendService.block(principal, request.userId()));
    }
    @DeleteMapping("/blocks/{userId}")
    public ApiResponse<Map<String, Object>> unblock(@AuthenticationPrincipal AuthenticatedUserPrincipal principal, @PathVariable Long userId) {
        return ApiResponse.success(friendService.unblock(principal, userId));
    }
    @DeleteMapping("/{userId}")
    public ApiResponse<Map<String, Object>> remove(@AuthenticationPrincipal AuthenticatedUserPrincipal principal, @PathVariable Long userId) {
        return ApiResponse.success(friendService.remove(principal, userId));
    }
}
